refactor(protocol): refactor tier provider & address caching and add ComposeVerifier

dantaik opened the pull request:
BEGIN_COMMIT_OVERRIDE
feat(protocol): add ComposeVerifier, TeeAnyVerifier, and ZkAnyVerifier
refactor(protocol): refactor TierProvider implementations
refactor(protocol): simplify mainnet address caching
END_COMMIT_OVERRIDE

dantaik committed:
refactor (#17951)

Co-authored-by: dantaik <dantaik@users.noreply.github.com>

dantaik committed:
temp revert

dantaik committed:
Merge branch 'main' into refactor_tiers

dantaik committed:
make LibRollupAddressCache backward compatible

dantaik committed:
rename

dantaik committed:
Revert "rename"

This reverts commit 8113502700cfb9841d65e79926f533a19c0ba105.

dantaik committed:
rename

dantaik committed:
AddressCache

dantaik committed:
AddressCache

dantaik committed:
more

dantaik committed:
forge fmt & update contract layout table

dantaik committed:
restore some changes

dantaik committed:
Update DevnetTierProvider.sol

dantaik committed:
f

dantaik committed:
Update TierProviderV2.sol

dantaik committed:
Update HeklaTierProvider.sol

dantaik committed:
Update TierProviderV2.sol

dantaik committed:
more

dantaik committed:
Update LibStrings.sol

dantaik committed:
Update LibStrings.sol

dantaik committed:
more

dantaik committed:
Update AddressResolver.t.sol

dantaik committed:
Update TierProviderBase.sol

dantaik committed:
feat(protocol): add ComposeVerifier (#17961)

dantaik committed:
Update AddressCache.sol

dantaik committed:
Update AddressCache.sol

dantaik committed:
Update ZkAnyVerifier.sol

dantaik committed:
Merge branch 'main' into refactor_tiers

dantaik committed:
Update ComposeVerifier.sol

adaki2004 commented on packages/protocol/contracts/L1/tiers/TierProviderBase.sol:
```diff
@@ -48,9 +48,20 @@ abstract contract TierProviderBase is ITierProvider {
         }
 
-        if (_tierId == LibTiers.TIER_SGX2) {
+        if (_tierId == LibTiers.TIER_TDX) {
```
It seems (since `TIER_TDX == 200` as well ass `TIER_SGX` and `TIER_TEE_ANY`) that one is enough, also their fields are the same. 
Is there any reason having all defined ?

dantaik replied:
With different names but same IDs, in TierProviderBase, we can configure all of them in the one single file so they have different verifiers.

adaki2004 commented on packages/protocol/contracts/verifiers/compose/ComposeVerifier.sol:
```diff
@@ -0,0 +1,88 @@
+
+            (bool success, bytes memory returnData) = subproofs[i].verifier.call(
+                abi.encodeCall(
+                    IVerifier.verifyProof,
```
Calling `SgxVerifier`'s `verifyProof()` - if we will start using the TEE any, we need to modify the code because SgxVerifier.sol line 146, it has a guard:
`onlyFromNamed(LibStrings.B_TAIKO)`.

Maybe best to mention it as a todo.

smtmfft replied:
It was initially added by Dani for security concern I guess @adaki2004? And yes, I think it can be removed now.

adaki2004 replied:
Im not sure if we can / shall remove it.

From mobile ATM but there was an issue allowing calling the contract by anybody.

I think it was the replay issue. So if caller is not the TaikoL1 and the call succeeds, it already exchanges the keys, so that the call from TaikoL1 womt suceed, this is why we need here. (If not mistaken).

smtmfft replied:
yes, but why a prover does that? it seems no harm for the others as he can not update others key, only negative impact to himself?

dantaik replied:
I read the code again, it seems we can indeed remove the check. Without the check, basically if someone has access to an instance's private-key, he can then remove the instance (and potentially add a new one). But this will not affect others as this instance will never be used by others.

adaki2004 replied:
> yes, but why a prover does that? it seems no harm for the others as he can not update others key, only negative impact to himself?

No, (afaik) it is not about that.

Imagine Bob is an honest (sgx) prover. Alice is a malicious actor who wants to trick Bob into losing money.
Alice monitors the mempool and when she sees the proving transaction, she extracts the proof and submit directly to the SgxVerifier contract with higher gas price.
Bob’s proof then would be revertes and possibly would loose money (bond). This guard is to prevent that situation.
(Basically it would work all the time and can increase proving time so that causes harm to alk the SGX prover to pay up to “infonity” just to prove a block and get back a bond).

it was also flagged  by SigmaPrime audit.

adaki2004 replied:
> I read the code again, it seems we can indeed remove the check. Without the check, basically if someone has access to an instance's private-key, he can then remove the instance (and potentially add a new one). But this will not affect others as this instance will never be used by others.

I'm referring to this audit finding.
<img width="971" alt="kép" src="https://github.com/user-attachments/assets/eb573c5c-ae64-4771-9aeb-bedb235b0038">

dantaik replied:
Can Alice frontruns Bob it means Bob shared his private key with Alice so they are considered the same party, right? Otherwise, how can Alice create a signature for Bob's instance?


```
      address oldInstance = ECDSA.recover(
            LibPublicInput.hashPublicInputs(
                _tran, address(this), newInstance, _ctx.prover, _ctx.metaHash, taikoChainId()
            ),
            _proof.data[24:]
        );
```

adaki2004 replied:
> Can Alice frontruns Bob it means Bob shared his private key with Alice so they are considered the same party, right? Otherwise, how can Alice create a signature for Bob's instance?
> 
> ```
>       address oldInstance = ECDSA.recover(
>             LibPublicInput.hashPublicInputs(
>                 _tran, address(this), newInstance, _ctx.prover, _ctx.metaHash, taikoChainId()
>             ),
>             _proof.data[24:]
>         );
> ```
Maybe i'm wrong or i dont see if we changed something in between, but here is my idea:

1. Bob has to re-create this "payload" off-chain, right ? (to sign it)
```
LibPublicInput.hashPublicInputs(
                _tran, address(this), newInstance, _ctx.prover, _ctx.metaHash, taikoChainId()
            ),
```
            
2. Alice cannot have access to the private key of Bob BUT she **sees** Bob's transaction **in the mempool**. She can extract this information then submit the smart contract call with higher gas tip. (Without that guard).

3. So Alice take advantage of the fact she can replace Bob's with the new instance, so that Bob cannot verify the transition with the old anymore.

Imo this creates the security vulnerability. (Please let me know if it is not or we mitigated somehow.. But i still think we need the `onlyFromNamed` ).

Brechtpd replied:
Because of the potential key switching it is indeed a problem and should be guarded.

dantaik replied:
> 2\. Alice cannot have access to the private key of Bob BUT she **sees** Bob's transaction **in the mempool**. She can extract this information then submit the smart contract call directly to SgxVerifier, with higher gas tip. (Without that guard).

I see.

adaki2004 commented on packages/protocol/contracts/verifiers/compose/TeeAnyVerifier.sol:
```diff
@@ -0,0 +1,24 @@
+import "./ComposeVerifier.sol";
+
+/// @title TeeAnyVerifier
+/// @notice This contract is a verifier for the Mainnet ZkVM that composes RiscZero and SP1
```
```suggestion
/// @notice This contract is a verifier for the Mainnet TEE proofs that composes SGX and TDX
```

dantaik committed:
Update packages/protocol/contracts/verifiers/compose/TeeAnyVerifier.sol

Co-authored-by: D <51912515+adaki2004@users.noreply.github.com>

dantaik committed:
Update TierProviderBase.sol

dantaik committed:
Update SgxVerifier.sol

dantaik committed:
Update TierProviderBase.sol

dantaik committed:
Update TierProviderBase.sol

dantaik committed:
Update SgxVerifier.t.sol

dantaik committed:
Update ITierProvider.sol

adaki2004 commented on packages/protocol/contracts/verifiers/SgxVerifier.sol:
```diff
@@ -149,7 +149,6 @@ contract SgxVerifier is EssentialContract, IVerifier {
     )
         external
-        onlyFromNamed(LibStrings.B_TAIKO)
     {
```
I still think it is a security issue as referred to here. https://github.com/taikoxyz/taiko-mono/pull/17960#discussion_r1728417427

Please see SigmaPrime report. 
<img width="971" alt="kép" src="https://github.com/user-attachments/assets/9f80b877-cb0a-4aaa-8159-4d2a85334890">

dantaik committed:
Merge branch 'main' into refactor_tiers

dantaik committed:
Update SgxVerifier.sol

dantaik committed:
Update Risc0Verifier.sol

dantaik committed:
delete

dantaik committed:
Update Risc0Verifier.sol

dantaik commented:
Based on the comments above:

- I deleted `ComposeVerifier.sol`.
- I changed the `verifyProof` function to **view** in `Risc0Verifier.sol` @smtmfft please verify.

dantaik commented on packages/protocol/contracts/verifiers/Risc0Verifier.sol:
```diff
@@ -53,6 +50,7 @@ contract RiscZeroVerifier is EssentialContract, IVerifier {
         TaikoData.TierProof calldata _proof
     )
         external
+        view
```
@smtmfft please take a look at this.

adaki2004 reviewed (commented):
Honestly, i liked the `ComposeVerifier` and the other derived contracts, which acted like routes. 
Beside the fact we had that guard issue with `SGX` (i think easily solveable by just guarding all the calls from: `TaikoL1 -> ComposeVerifier -> Verifier` especially in SGX case), i think it was a good way to achieve what Yue wanted with that distribution diagram.
But let me copy the diagram again here, so we have that:
![kép](https://github.com/user-attachments/assets/538b261f-f4fc-43f8-b998-f09e55ccd5d8)

Also this draft PR trying to achieve that too - not that dynamic like the `ComposeVerifier` tho -  but more lightweight and gas efficient, if we wish to proceed further with that, leaving it here: https://github.com/taikoxyz/taiko-mono/pull/17894

dantaik committed:
Revert "delete"

This reverts commit f206780c2a25195c813528b42ebffd2514a7642e.

dantaik commented:
> Also this draft PR trying to achieve that too - not that dynamic like the `ComposeVerifier` tho - but more lightweight and gas efficient, if we wish to proceed further with that, leaving it here: #17894

@adaki2004  added back those contracts, and made some changes to the permission check in ComposeVerifier and SgxVerifier. Please take another look.

dantaik committed:
Merge branch 'main' into refactor_tiers

dantaik committed:
Update ComposeVerifier.sol

dantaik committed:
fix

dantaik committed:
fix

dantaik committed:
Update ComposeVerifier.sol

adaki2004 commented on packages/protocol/contracts/verifiers/compose/ComposeVerifier.sol:
```diff
@@ -0,0 +1,101 @@
+    )
+        external
+    {
+        if (!isCallerAuthorized(msg.sender)) revert INVALID_CALLER();
```
Hmm, why not `onlyFromNamed(LibStrings.B_TAIKO)` ?

dantaik replied:
isCallerAuthorized is to make testing easier.

Brechtpd commented on packages/protocol/contracts/verifiers/compose/ComposeVerifier.sol:
```diff
@@ -0,0 +1,101 @@
+            if (!verifierFound) revert INVALID_VERIFIER();
+
+            (bool success, bytes memory returnData) = subproofs[i].verifier.call(
+                abi.encodeCall(
+                    IVerifier.verifyProof,
```
what was the reason again to not just call verifyProof directly on the address?

dantaik replied:
Using the low level call may be an overkill here. I refactored the code to require the number of sub-proofs much match the threshold and each sub-proof must be verifiable.

dantaik committed:
Update ComposeVerifier.sol

dantaik committed:
Revert "Update ComposeVerifier.sol"

This reverts commit 8d9b9ceb2a8c1455281402dad504ddb3d17ae1c8.

dantaik committed:
more

dantaik committed:
Merge branch 'main' into refactor_tiers

dantaik committed:
more

dantaik committed:
more

dantaik committed:
more

dantaik committed:
forge fmt & update contract layout table

dantaik committed:
Update ComposeVerifeir.t.sol

dantaik committed:
Update ComposeVerifier.sol

dantaik committed:
Update ComposeVerifeir.t.sol

dantaik committed:
fix

adaki2004 committed:
Merge branch 'main' into refactor_tiers

dantaik commented on packages/protocol/contracts/L1/tiers/ITierProvider.sol:
```diff
@@ -24,6 +24,7 @@ interface ITierProvider {
     /// @dev Retrieves the IDs of all supported tiers.
     /// Note that the core protocol requires the number of tiers to be smaller
     /// than 256. In reality, this number should be much smaller.
+    /// Additionally, each tier's ID value must be unique.
```
```suggestion
```

dantaik committed:
Update packages/protocol/contracts/L1/tiers/ITierProvider.sol

dantaik committed:
more

dantaik committed:
Merge branch 'main' into refactor_tiers

dantaik committed:
Update RollupAddressCache.sol
